src/ILCompiler.Compiler/src/CppCodeGen/CppWriter.cs

```diff
@@ -729,192 +733,204 @@ private String GetCodeForVirtualMethod(MethodDesc method, int slot)
-            {
-                sb.AppendLine();
-                sb.Append("{");
-                sb.Indent();
-                sb.AppendLine();
-                sb.Append("sizeof(");
-                sb.Append(GetCppSignatureTypeName(((ArrayType)type).ElementType)); // EEType::_usComponentSize
-                sb.Append("),");
-                sb.AppendLine();
-                sb.Append("0x");
-                sb.Append(flags.ToStringInvariant("x4"));  // EEType::_usFlags
-                sb.Append(",");
-                sb.AppendLine();
-                sb.Append("3 * sizeof(void*),"); // EEType::_uBaseSize
-            }
-            else
-            if (type.IsArray)
-            {
-                sb.AppendLine();
-                sb.Append("{");
-                sb.Indent();
-                sb.AppendLine();
-                sb.Append("sizeof(");
-                sb.Append(GetCppSignatureTypeName(((ArrayType)type).ElementType)); // EEType::_usComponentSize
-                sb.Append("),");
-                sb.AppendLine();
-                sb.Append("0x");
-                sb.Append(flags.ToStringInvariant("x4"));  // EEType::_usFlags
-                sb.Append(",");
-                sb.AppendLine();
-                sb.Append("3 * sizeof(void*) + ");                            // EEType::_uBaseSize
-                sb.Append(((ArrayType)type).Rank.ToStringInvariant());
-                sb.Append("* sizeof(int32_t) * 2,");
-            }
-            else
-            {
-                // sizeof(void*) == size of object header
-                sb.AppendLine();
-                sb.Append("{");
-                sb.Indent();
-                sb.AppendLine();
-                sb.Append("0,");
-                sb.AppendLine();
-                sb.Append("0x");                                           // EEType::_usComponentSize
-                sb.Append(flags.ToStringInvariant("x"));   // EEType::_usFlags
-                sb.Append(",");
-                sb.AppendLine();
-                sb.Append("AlignBaseSize(sizeof(void*)+sizeof(");             // EEType::_uBaseSize
-                sb.Append(GetCppTypeName(type));
-                sb.Append(")),");
-            }
 
-            sb.AppendLine();
+        }
+        private String GetCodeForNodeData(List<Tuple<bool, int>> nodeDataDivs, Relocation[] relocs, byte[] byteData, DependencyNode node)
+        {
+            CppGenerationBuffer nodeDataDecl = new CppGenerationBuffer();
+            int relocCounter = 0;
+            int divisionStartIndex = 0;
+            nodeDataDecl.Indent();
+            nodeDataDecl.AppendLine();
 
-            // base type
-            if (type.IsArray)
+            for (int i = 0; i < nodeDataDivs.Count; i++)
             {
-                sb.Append(GetCppMethodDeclarationName(((ArrayType)type).ElementType, "__getMethodTable"));
-                sb.Append("()");
+                nodeDataDecl.Indent();
+
+                if (nodeDataDivs[i].Item1)
```

Comment:
More of a question to @nattress, but would handling `IEETypeNode` and `IMethodCodeNode` instead of the concrete type make this code path multi-file mode friendly?